### ISSUE TITLE
Image shifts on fullscreen with dual monitor setup

### ISSUE THREAD

**Issue by Netzwerk2** [woelper/oculante#527]
**Describe the bug**
When I enter fullscreen (on my right monitor) the image shifts to the right. Exiting fullscreen restores the image to its original position.
I also tried changing the primary monitor, but that didn't do anything.

**To Reproduce**
1. Open an image on the right monitor.
2. Enter fullscreen

**Expected behavior**
The image position should stay the same.

**Desktop**
Arch Linux, Xfce (X11)

**Additional context**
Commenting out the following code block in `utils::toggle_fullscreen` fixes the issue. I would have opened a PR to remove it, but I'm not sure why that code is even there in the first place :)

```rs
    if !fullscreen {
        let mut window_pos = app.window().position();
        window_pos.1 += 40;

        debug!("Not fullscreen. Storing offset: {:?}", window_pos);

        let dpi = app.window().dpi();
        debug!("{:?}", dpi);
        window_pos.0 = (window_pos.0 as f64 / dpi) as i32;
        window_pos.1 = (window_pos.1 as f64 / dpi) as i32;
        #[cfg(target_os = "macos")]
        {
            // tweak for osx titlebars
            window_pos.1 += 8;
        }

        // if going from window to fullscreen, offset by window pos
        state.image_geometry.offset.x += window_pos.0 as f32;
        state.image_geometry.offset.y += window_pos.1 as f32;

        // save old window pos
        state.fullscreen_offset = Some(window_pos);
    } else if let Some(sf) = state.fullscreen_offset {
        state.image_geometry.offset.x -= sf.0 as f32;
        state.image_geometry.offset.y -= sf.1 as f32;
    }
```

**Comment by woelper** [comment]
Thank you for the report!

I think that may have been something that changed in `winit`. Some time ago, there was a huge difference in image position once you went into full and back. The goal was that the image position does not change (the pixel under the cursor stays the exact same) whether you enter fullscreen or not. 

I'll have a look at removing the code, but I need to test this on mac/windows and on high dpi displays first.

**Comment by woelper** [comment]
I've tested on Ubuntu and Mac, and I see a slight vertical offset, but I can't reproduce any shifting in horizontal direction.

**Comment by Netzwerk2** [comment]
I tested with GNOME and even Penrose as WMs and it happened with both for me. I can send a Video later.

**Comment by woelper** [comment]
Oh, that is odd! I'll see if I can reproduce this on other linux desktop environments or distributions.

**Comment by Netzwerk2** [comment]
I also tested this on Ubuntu 24.04.1 LTS with xmonad (X11) as WM. Still the same issue.

<details>
<summary>Here's the output of <code class="notranslate">xrandr -q</code> (Ubuntu, xmonad)</summary>

```
Screen 0: minimum 8 x 8, current 4160 x 1440, maximum 32767 x 32767
DVI-D-0 connected 1600x1200+2560+0 (normal left inverted right x axis y axis) 408mm x 306mm
   1600x1200     60.00*+
   1400x1050     74.87    59.98  
   1280x1024     75.02    60.02  
   1280x960      60.00  
   1152x864      75.00  
   1024x768      75.03    70.07    60.00  
   800x600       75.00    72.19    60.32    56.25  
   640x480       75.00    72.81    59.94  
HDMI-0 connected primary 2560x1440+0+0 (normal left inverted right x axis y axis) 708mm x 399mm
   2560x1440     59.95*+
   1920x1080     60.00    59.94    50.00    29.97    25.00    23.98    60.00    50.04  
   1680x1050     59.95  
   1600x900      60.00  
   1280x1024     75.02    60.02  
   1280x800      59.81  
   1280x720      60.00    59.94    50.00  
   1024x768      75.03    60.00  
   800x600       75.00    60.32  
   720x576       50.00  
   720x480       59.94  
   640x480       75.00    59.94    59.93  
DP-0 disconnected (normal left inverted right x axis y axis)
DP-1 disconnected (normal left inverted right x axis y axis)
```

</details>

<details>
<summary>And for my system (Arch, Xfce)</summary>

```
Screen 0: minimum 8 x 8, current 3840 x 1080, maximum 32767 x 32767
DP-0 disconnected (normal left inverted right x axis y axis)
DP-1 connected 1920x1080+0+0 (normal left inverted right x axis y axis) 527mm x 296mm
   1920x1080     60.00*+  59.94    50.00
   1680x1050     59.95
   1600x900      60.00
   1440x900      59.89
   1280x1024     75.02    60.02
   1280x800      59.81
   1280x720      60.00    59.94    50.00
   1152x864      75.00
   1024x768      75.03    70.07    60.00
   800x600       75.00    72.19    60.32    56.25
   720x576       50.00
   720x480       59.94
   640x480       75.00    72.81    59.94
DP-2 disconnected (normal left inverted right x axis y axis)
DP-3 disconnected (normal left inverted right x axis y axis)
HDMI-0 connected primary 1920x1080+1920+0 (normal left inverted right x axis y axis) 480mm x 270mm
   1920x1080     60.00*+  74.97    59.94    50.00
   1680x1050     59.95
   1600x900      60.00
   1440x900      59.89
   1400x1050     59.98
   1280x1024     75.02    60.02
   1280x800      59.81
   1280x720      60.00    59.94    50.00
   1152x864      75.00
   1024x768      75.03    60.00
   800x600       75.00    60.32
   720x576       50.00
   720x480       59.94
   640x480       75.00    59.94    59.93
DP-4 disconnected (normal left inverted right x axis y axis)
DP-5 disconnected (normal left inverted right x axis y axis)
```

</details>

**Comment by Stoppedpuma** [comment]
I can't reproduce this under Sway on Arch, perhaps this is an issue with x11 systems? @Netzwerk2 Are you able to reproduce this under Wayland?

**Comment by woelper** [comment]
I was able to reproduce this on a fresh Fedora VM today, but only if the window is rather small and at the right side of the  screen. We definitely need to test any fix on different operating systems and desktop environments.

**Comment by Netzwerk2** [comment]
I tested it on GNOME with Wayland. While it didn't jump that much, there still was a noticeable difference in position between fullscreen on and off.
> The goal was that the image position does not change (the pixel under the cursor stays the exact same) whether you enter fullscreen or not.

This was definitely not the case. I can't record a video though, cause for some reason OBS misses the options to record a screen/window on Wayland.

**Comment by Netzwerk2** [comment]
I also just noticed an issue with the picking preview. When I enter fullscreen and don't move the mouse, the preview is at a slightly wrong position.

Should I open a second issue for that?

https://github.com/user-attachments/assets/730a58cf-3035-43f8-ac39-17658ed36d64

**Comment by Stoppedpuma** [comment]
> I was able to reproduce this on a fresh Fedora VM today, but only if the window is rather small and at the right side of the screen. We definitely need to test any fix on different operating systems and desktop environments.

I think I was able to reproduce this with fit on resize turned on? Maybe not the same thing and likely intentional but it seems to just go to the top left of the screen when I follow your instructions. I can't seem to reproduce this at all with fit to window size turned on.

![image](https://github.com/user-attachments/assets/ef648b97-10bd-47ce-b7b7-9f289a5400b9)
![image](https://github.com/user-attachments/assets/c2d71373-d6f5-4c96-86ee-f361b240b8ad)

> I also just noticed an issue with the picking preview. When I enter fullscreen and don't move the mouse, the preview is at a slightly wrong position.
Should I open a second issue for that?

Can you see if you can reproduce this on the `dynamicimage` branch? The picker has changed on that branch.

**Comment by Netzwerk2** [comment]
> Can you see if you can reproduce this on the `dynamicimage` branch? The picker has changed on that branch.

Yes, the issue is still there.

**Comment by woelper** [comment]
What is strange is that I can reproduce the issue, but commenting out the code does not fix it. Have you checked moving the window around the screen and going into fullscreen? I have a hard time imagining how this should work with that code commented and moving the window somewhere, panning the image and the same pixel should still be under the cursor if you enter fullscreen.

**Comment by Netzwerk2** [comment]
You're right. If I comment the code out and scale the window down, the image doesn't remain at the same place when entering/exiting fullscreen. But it isn't moved completely out of view, unlike the original code.
